src/categorical_algebra/StructuredCospans.jl

```diff
@@ -0,0 +1,271 @@
+using ...Theories: Category, CatDesc, AttrDesc
+import ...Theories: dom, codom, compose, ⋅, id
+
+# Generic structured cospans
+############################
+
+""" Structured multicospan.
+
+A structured multicospan is like a structured cospan except that it may have a
+number of legs different than two.
+
+See also: [`StructuredCospan`](@ref).
+"""
+@auto_hash_equals struct StructuredMulticospan{L, Cosp <: Multicospan,
+                                               Feet <: AbstractVector}
+  cospan::Cosp
+  feet::Feet
+
+  """ Construct structured multicospan in L-form.
+  """
+  StructuredMulticospan{L}(cospan::Cosp, feet::Feet) where
+      {L, Cosp <: Multicospan, Feet <: AbstractVector} =
+    new{L,Cosp,Feet}(cospan, feet)
+end
+
+""" Construct structured multicospan in R-form.
+"""
+function StructuredMulticospan{L}(x, cospan::Multicospan) where L
+  StructuredMulticospan{L}(
+    Multicospan(x, map(leg -> shift_left(L, x, leg), legs(cospan))),
+    feet(cospan))
+end
+
+apex(cospan::StructuredMulticospan) = apex(cospan.cospan)
+legs(cospan::StructuredMulticospan) = legs(cospan.cospan)
+feet(cospan::StructuredMulticospan) = cospan.feet
+
+""" Structured cospan.
+
+The first type parameter `L` encodes a functor L: A → X from the base category
+`A`, often FinSet, to a category `X` with "extra structure." An L-structured
+cospan is then a cospan in X whose feet are images under L of objects in A. The
+category X is assumed to have pushouts.
+
+Structured cospans form a double category with no further assumptions on the
+functor L. To obtain a symmetric monoidal double category, L must preserve
+finite coproducts. In practice, L usually has a right adjoint R: X → A, which
+implies that L preserves all finite colimits. It also allows structured cospans
+to be constructed more conveniently from an object x in X plus a cospan in A
+with apex R(x).
+
+See also: [`StructuredMulticospan`](@ref).
+"""
+const StructuredCospan{L, Cosp <: Cospan, Feet <: StaticVector{2}} =
+  StructuredMulticospan{L,Cosp,Feet}
+
+""" Construct structured cospan in L-form.
+"""
+StructuredCospan{L}(cospan::Cospan, feet::StaticVector{2}) where L =
+  StructuredMulticospan{L}(cospan, feet)
+
+""" Construct structured cospan in R-form.
+"""
+StructuredCospan{L}(apex, cospan::Cospan) where L =
+  StructuredMulticospan{L}(apex, cospan)
+
+left(cospan::StructuredCospan) = first(legs(cospan))
+right(cospan::StructuredCospan) = last(legs(cospan))
+
+# Category of structured cospans
+################################
+
+""" Object in the category of L-structured cospans.
+"""
+@auto_hash_equals struct StructuredCospanOb{L,T}
+  ob::T
+  StructuredCospanOb{L}(ob::T) where {L,T} = new{L,T}(ob)
+end
+
+function StructuredCospan(cospan::Cospan, lfoot::StructuredCospanOb{L},
+                          rfoot::StructuredCospanOb{L}) where L
+  StructuredCospan{L}(cospan, SVector(lfoot.ob, rfoot.ob))
+end
+
+@instance Category{StructuredCospanOb, StructuredCospan} begin
```

Comment:
Should this be DoubleCategory once #278 lands?

Reply:
I was planning to first do hypergraph category, for feature parity with what's already in AlgebraicPetri, but yes, eventually symmetric monoidal double category!